pandas/tests/indexes/multi/test_constructor.py

```diff
@@ -0,0 +1,434 @@
+
+    idx1 = []
+    idx2 = ['a', 'b']
+    tm.assert_raises_regex(ValueError, '^all arrays must '
+                           'be same length$',
+                           MultiIndex.from_arrays, [idx1, idx2])
+
+    idx1 = [1, 2, 3]
+    idx2 = []
+    tm.assert_raises_regex(ValueError, '^all arrays must '
+                           'be same length$',
+                           MultiIndex.from_arrays, [idx1, idx2])
+
+
+def test_from_tuples():
+    tm.assert_raises_regex(TypeError, 'Cannot infer number of levels '
+                           'from empty list',
+                           MultiIndex.from_tuples, [])
+
+    expected = MultiIndex(levels=[[1, 3], [2, 4]],
+                          labels=[[0, 1], [0, 1]],
+                          names=['a', 'b'])
+
+    # input tuples
+    result = MultiIndex.from_tuples(((1, 2), (3, 4)), names=['a', 'b'])
+    tm.assert_index_equal(result, expected)
+
+
+def test_from_tuples_iterator():
+    # GH 18434
+    # input iterator for tuples
+    expected = MultiIndex(levels=[[1, 3], [2, 4]],
+                          labels=[[0, 1], [0, 1]],
+                          names=['a', 'b'])
+
+    result = MultiIndex.from_tuples(zip([1, 3], [2, 4]), names=['a', 'b'])
+    tm.assert_index_equal(result, expected)
+
+    # input non-iterables
+    with tm.assert_raises_regex(
+            TypeError, 'Input must be a list / sequence of tuple-likes.'):
+        MultiIndex.from_tuples(0)
+
+
+def test_from_tuples_empty():
+    # GH 16777
+    result = MultiIndex.from_tuples([], names=['a', 'b'])
+    expected = MultiIndex.from_arrays(arrays=[[], []],
+                                      names=['a', 'b'])
+    tm.assert_index_equal(result, expected)
+
+
+def test_from_tuples_index_values(idx):
+    result = MultiIndex.from_tuples(idx)
+    assert (result.values == idx.values).all()
+
+
+def test_from_product_empty():
+    # 0 levels
+    with tm.assert_raises_regex(
+            ValueError, "Must pass non-zero number of levels/labels"):
+        MultiIndex.from_product([])
+
+    # 1 level
+    result = MultiIndex.from_product([[]], names=['A'])
+    expected = pd.Index([], name='A')
+    tm.assert_index_equal(result.levels[0], expected)
+
+    # 2 levels
+    l1 = [[], ['foo', 'bar', 'baz'], []]
+    l2 = [[], [], ['a', 'b', 'c']]
+    names = ['A', 'B']
+    for first, second in zip(l1, l2):
+        result = MultiIndex.from_product([first, second], names=names)
+        expected = MultiIndex(levels=[first, second],
+                              labels=[[], []], names=names)
+        tm.assert_index_equal(result, expected)
+
+    # GH12258
+    names = ['A', 'B', 'C']
+    for N in range(4):
+        lvl2 = lrange(N)
+        result = MultiIndex.from_product([[], lvl2, []], names=names)
+        expected = MultiIndex(levels=[[], lvl2, []],
+                              labels=[[], [], []], names=names)
+        tm.assert_index_equal(result, expected)
+
+
```

Comment:
parametrize

Reply:
Broke into several parameterized tests as the length of the names and the levels were different in each.  Trying to do it all in one level seemed clunky to me.